### PR TITLE
feat: add spare table data structure

### PR DESCRIPTION
**Pull request by manncodes**
#### Description of Change

Added implementation of sparse table for any 'duplicate invariant function'. Implementation of min(a1,a2...,aN) being the duplicate invariant function is done. Sparse tables are useful data structures to answer ranged queries in O(1), with precomputation of O(N*log(N)).

Implementation in `sparse_table.cpp` contains insightful comments and abides the Doxygen standards.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
The implementation function can be 'hot swapped' by changing just 2 lines of code, so the implementation of sparse table for min() queries becomes sparse table for gcd().

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1502"></a>

### CODE REVIEW

**Comment by manncodes** [comment]
@Panquesito7 can you review it please?

File: data_structures/sparse_table.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
+#include <array>     /// for std::array
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:3]: This should be added below, the algorithm description.
```suggestion
```

File: data_structures/sparse_table.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+ * data structure is, that it can only be used on immutable arrays. This means,
+ * that the array cannot be changed between two queries.
+ *
+ * If any element in the array changes, the complete data structure has to be
+ * recomputed.
+ *
+ * @warning this sparse table is made for min(a1,a2,...an) duplicate invariant
+ * function. This implementation can be changed to other functions like
+ * gcd(),lcm(),max() by changing few lines of code.
+ */
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:25]:
```suggestion
 */
 
#include <array>     /// for std::array
#include <cassert>   /// for assert
#include <iostream>  /// for IO operations
```

File: data_structures/sparse_table.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
+#include <array>     /// for std::array
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+/**
+ *
+ * @file
+ * @brief Implementation of Sparse Table
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:7]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: data_structures/sparse_table.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+ *
+ * @author [manncodes](https://github.com/manncodes)
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:9]:
```suggestion
 * @author [Mann Patel](https://github.com/manncodes)
```

File: data_structures/sparse_table.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+ * It can answer most range queries in O(logn), but its true power is answering
+ * range minimum queries (or equivalent range maximum queries). For those
+ * queries it can compute the answer in O(1) time. The only drawback of this
+ * data structure is, that it can only be used on immutable arrays. This means,
+ * that the array cannot be changed between two queries.
+ *
+ * If any element in the array changes, the complete data structure has to be
+ * recomputed.
+ *
+ * @warning this sparse table is made for min(a1,a2,...an) duplicate invariant
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:22]:
```suggestion
 * @warning this sparse table is made for `min(a1,a2,...an)` duplicate invariant
```

File: data_structures/sparse_table.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+ * function. This implementation can be changed to other functions like
+ * gcd(),lcm(),max() by changing few lines of code.
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:24]:
```suggestion
 * `gcd()`, `lcm()`, and `max()` by changing a few lines of code.
```

File: data_structures/sparse_table.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+/**
+ * @namespace sparse_table
+ * @brief Implementation of Sparse Table for function : min()
+ *
+ */
+namespace sparse_table {
+
+/**
+ * @brief A struct to represent sparse table
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:41]: Please provide a better description of this structure.

File: data_structures/sparse_table.cpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    int n;
+    std::array<int, N> a;
+
+    //
+    // Sparse table related variables.
+    //
+    // ST[j][i] : the sparse table value (min, max, ...etc) in the interval [i,
+    // i + (2^j) - 1]. LOG[i]   : floor(log2(i)).
+    std::array<std::array<int, N>, M> ST;
+    std::array<int, N> LOG;
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:59]: Add a one-line description of these variables (see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program for more information).

File: data_structures/sparse_table.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        int y = ST[g][r - (1 << g) + 1];
+        return (a[x] <= a[y] ? x : y);
+    }
+};
+}  // namespace sparse_table
+}  // namespace data_structures
+
+/**
+ * @brief testcase for sparse_table
+ */
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:111]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
```

File: data_structures/sparse_table.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+/**
+ * @brief Main function
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, char *argv[]) {
+    test();  // execute the tests
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:139]:
```suggestion
    test();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work.

**Comment by manncodes** [comment]
@Panquesito7 done with changes! 👍

File: data_structures/sparse_table.cpp [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+/**
+ * @namespace data_structures
+ * @brief Data Structures algorithms
+ */
+namespace data_structures {
+
+/**
+ * @namespace sparse_table
+ * @brief sparse table and it's util functions.
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:35]:
```suggestion
 * @brief Functions for Implementation of [Sparse Table](https://brilliant.org/wiki/sparse-table/)
```

File: data_structures/sparse_table.cpp [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+ */
+namespace sparse_table {
+
+/**
+ * @brief A struct to represent sparse table for `min()` as their invariant
+ * function, for the given array `A`. The answer to queries are stored in array
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:41]:
```suggestion
 * function, for the given array `A`. The answer to queries are stored in the array
```

File: data_structures/sparse_table.cpp [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+ * ST.
+ */
+struct Sparse_table {
+    const static int N = 12345;  ///< the maximum size of the array.
+    const static int M = 14;     ///< ceil(log2(N)).
+    int n;                       ///< size of input array.
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:47]: It's always best to initialize variables to `0`, as they can display wrong values in some cases.
```suggestion
    int n = 0;                       ///< size of input array.
```

File: data_structures/sparse_table.cpp [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+ */
+namespace sparse_table {
+
+/**
+ * @brief A struct to represent sparse table for `min()` as their invariant
+ * function, for the given array `A`. The answer to queries are stored in array
+ * ST.
+ */
+struct Sparse_table {
+    const static int N = 12345;  ///< the maximum size of the array.
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:45]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/sparse_table.cpp (hunk 13 above)
Comment (manncodes) [reply data_structures/sparse_table.cpp:47]: should I initialize other arrays to zero to?

File: data_structures/sparse_table.cpp (hunk 14 above)
Comment (manncodes) [reply data_structures/sparse_table.cpp:45]: > Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

would you recommend using `size_t` ?

File: data_structures/sparse_table.cpp [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+    st.n = testcase_size;
+
+    st.buildST();  ///< precomputing sparse tree
+
+    // pass queries of the form: [l,r]
+    assert(st.query(1, 9) == 1);  ///< as 1 is smallest from 1..9
+    assert(st.query(2, 6) == 2);  ///< as 2 is smallest from 2..6
+    assert(st.query(3, 8) == 3);  ///< as 3 is smallest from 3..8
+
+    std::cout << "Testcase passed!" << std::endl;
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:123]:
```suggestion
    std::cout << "Self-test implementations passed!" << std::endl;
```

**Review by Panquesito7 — DISMISSED** [review]
Amazing work; I like how you've made & improved the code so well. 😄
Code and documentation look good and very refined; LGTM. 👍 🎉

File: data_structures/sparse_table.cpp [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+ * Table](https://brilliant.org/wiki/sparse-table/)
+ */
+namespace sparse_table {
+
+/**
+ * @brief A struct to represent sparse table for `min()` as their invariant
+ * function, for the given array `A`. The answer to queries are stored in the
+ * array ST.
+ */
+const static uint32_t N = 12345;  ///< the maximum size of the array.
```
Comment (ayaankhan98) [inline data_structures/sparse_table.cpp:45]:
```suggestion
constexpr uint32_t N = 12345;  ///< the maximum size of the array.
```

File: data_structures/sparse_table.cpp [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+const static uint8_t M = 14;      ///< ceil(log2(N)).
```
Comment (ayaankhan98) [inline data_structures/sparse_table.cpp:46]:
```suggestion
constexpr uint8_t M = 14;      ///< ceil(log2(N)).
```

**Review by ayaankhan98 — COMMENTED** [review]
Rest looks good to me.

**Review by ayaankhan98 — COMMENTED** [review]
```
test.cpp: In member function ‘void data_structures::sparse_table::Sparse_table::buildST()’:
test.cpp:74:37: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   74 |         for (size_t j = 1; (1 << j) <= n; ++j) {
      |                            ~~~~~~~~~^~~~
cc1plus: all warnings being treated as errors

shell returned 1
```

**Comment by manncodes** [comment]
> ```
> test.cpp: In member function ‘void data_structures::sparse_table::Sparse_table::buildST()’:
> test.cpp:74:37: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
>    74 |         for (size_t j = 1; (1 << j) <= n; ++j) {
>       |                            ~~~~~~~~~^~~~
> cc1plus: all warnings being treated as errors
> 
> shell returned 1
> ```

should I just static cast the right shifted numbers `1<<j` to `size_t` ?

**Comment by manncodes** [comment]
compiled_capture
I see no compilation errors, when ran on gitpod ✔️

**Comment by manncodes** [comment]
@ayaankhan98  @Panquesito7 Can you review it please? Thanks!

**Review by ayaankhan98 — DISMISSED** [review]

File: data_structures/sparse_table.cpp [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+
+struct Sparse_table {
+    size_t n = 0;  ///< size of input array.
+
+    /** @warning check if `N` is not less than `n`. if so, manually increase the
+     * value of N */
+
+    std::array<int64_t, N> A = {};  ///< input array to perform RMQ.
+    std::array<std::array<int64_t, N>, M> ST =
+        {};  ///< the sparse table storing `min()` values for given interval.
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:56]: You can also use `{}` without an `=` which will make the code shorter.
```suggestion
    std::array<std::array<int64_t, N>, M> ST{};  ///< the sparse table storing `min()` values for given interval.
```

File: data_structures/sparse_table.cpp [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+     * (i.e. from l to r inclusive).
+     * @param l the left index of the range (inclusive).
+     * @param r the right index of the range (inclusive).
+     * @return the computed value of the given interval.
+     * @complexity: O(1)
+     */
+    int64_t query(int64_t l, int64_t r) {
+        int64_t g = LOG[r - l + 1];  ///< smallest power of 2 covering [l,r]
+        int64_t x = ST[g][l];
+        int64_t y = ST[g][r - (1 << g) + 1];
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:95]: Missing one-line documentation/description (see the line above).

File: data_structures/sparse_table.cpp [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    std::array<int64_t, 10> testcase = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+    size_t testcase_size = sizeof(testcase) / sizeof(testcase[0]);
+
+    data_structures::sparse_table::Sparse_table st{};
```
Comment (Panquesito7) [inline data_structures/sparse_table.cpp:110]: Maybe provide a description? The more description you provide, the better understanding for the readers.

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

File: data_structures/sparse_table.cpp (hunk 18 above)
Comment (manncodes) [reply data_structures/sparse_table.cpp:56]: I don't mind that, `clang-tidy` suggested that though!😄

**Review by Panquesito7 — APPROVED** [review]
Great work! Thanks!

**Comment by Panquesito7** [comment]
@manncodes, join our [Discord](https://discord.gg/c7MnfGFGa6) server to stay tuned about new announcements, ask for help, and receive answers quickly, discuss and talk with other developers, and much more. Join if you haven't already. Thanks! 😄

**Pull request merged**